src/lashup_kv.erl

```diff
@@ -229,72 +232,112 @@ init_db(Nodes) ->
   mnesia:change_table_copy_type (schema, node(), disc_copies), % If the node was already running
   {ok, _} = application:ensure_all_started(mnesia),
   ExistingTables = mnesia:system_info(tables),
-  Tables = [kv],
+  Tables = [?KV_TABLE, nclock],
   TablesToCreate = Tables -- ExistingTables,
+  Alltables = TablesToCreate ++ ExistingTables,
   lists:foreach(fun create_table/1, TablesToCreate),
-  ok = mnesia:wait_for_tables(Tables, infinity).
+  ok = mnesia:wait_for_tables(Alltables, infinity),
+  ok = maybe_upgrade_table(ExistingTables).
 
-create_table(kv) ->
-  {atomic, ok} =  mnesia:create_table(kv, [
-    {attributes, record_info(fields, kv)},
+create_table(Table) ->
+  {atomic, ok} =  mnesia:create_table(Table, [
+    {attributes, get_record_info(Table)},
     {disc_copies, [node()]},
     {type, set}
   ]).
 
+get_record_info(kv2) ->
+  record_info(fields, kv2);
+get_record_info(nclock) ->
+  record_info(fields, nclock).
 
--spec(mk_write_fun(Key :: key(), OldVClock :: riak_dt_vclock:vclock() | undefined, Op :: riak_dt_map:map_op())
-      -> (fun())).
+maybe_upgrade_table(ExistingTables) ->
+  case {lists:member(?KV_TABLE, ExistingTables), lists:member(?OLD_KV_TABLE, ExistingTables)} of
+    {true, _} ->
+        ok; %% already upgraded
+    {false, false} ->
+        ok; %% nothing to upgrade
+    {false, true} ->
+        upgrade_table(kv)
+  end.
+
+upgrade_table(kv) ->
+  F = fun() ->
+        Records = mnesia:select(?OLD_KV_TABLE, [{'_', [], ['$_']}]),
```

Comment:
Rather than fetching all the records (on the heap), and writing them all.
If you want to use foldl, since you're basically selecting across the entire old table, maybe use this:
http://erlang.org/doc/man/mnesia.html#foldl-3

But since you're not accumulating, you can just do a write in your callback function.